pkg/cli/mirror/util.go

```diff
@@ -113,3 +128,326 @@ func (o *MirrorOptions) checkErr(err error, acceptableErr func(error) bool, logM
 
 	return nil
 }
+
+func getPlatformKey(typeWithPlatorm interface{}) (*OperatorCatalogPlatform, error) {
+	if typeWithPlatorm == nil {
+		return nil, errors.New("no manifest or config provided, so unable to generate a platform key")
+	}
+	switch v := typeWithPlatorm.(type) {
+	case *v1.Platform:
+		return &OperatorCatalogPlatform{
+			os:           v.OS,
+			architecture: v.Architecture,
+			variant:      v.Variant,
+			isIndex:      true,
+		}, nil
+	case *v1.Descriptor:
+		if v.Platform == nil {
+			return nil, errors.New("no platform provided in descriptor, so unable to generate a platform key")
+		}
+		return &OperatorCatalogPlatform{
+			os:           v.Platform.OS,
+			architecture: v.Platform.Architecture,
+			variant:      v.Platform.Variant,
+			isIndex:      true,
+		}, nil
+	case *v1.ConfigFile:
+		// ConfigFile only comes into play when an image is a single architecture catalog.
+		// Callers may need to override the isIndex field if necessary
+		return &OperatorCatalogPlatform{
+			os:           v.OS,
+			architecture: v.Architecture,
+			variant:      v.Variant,
+			isIndex:      false,
+		}, nil
+	default:
+		// should not happen... this is a programmer error
+		return nil, fmt.Errorf("expected a manifest or config, but %T provided, so unable to generate a platform key", v)
+	}
+}
+
+/*
+getImageDigests will fetch one or more image digests for a given image reference.
+This function handles both "manifest list" and "image" references. A "manifest list"
+can return one or more digest references, whereas a "image" reference only returns a single
+digest reference.
+
+# Arguments
+
+• ctx: A cancellation context
+
+• imageRef: a docker image reference which will be either remotely accessed (if necessary)
+and used as a basis for the image references returned from this function
+
+• layoutPath: an optional OCI layout path. If provided imageRef will NOT be fetched remotely.
+This OCI layout path could refer to a "manifest list" or an "image".
+
+• insecure: flag to indicate if plain http is used or when skipping TLS verification
+
+# Returns
+
+• map[OperatorCatalogPlatform][]name.Digest: If no error occurs, returns a map whose key is Platform
+and value is a slice of one or more image digest references for that architecture.
+If an error occurs the map will always be initialized (i.e non-nil) but could have partial results.
+
+• error: non-nil if an error occurs, nil otherwise
+*/
+func getImageDigests(ctx context.Context, imageRef string, layoutPath *layout.Path, insecure bool) (map[OperatorCatalogPlatform][]name.Reference, error) {
```

Comment:
Is it a possibility to return a map using `imageInfo` (like in catalog_images), so that, instead of keeping just the digest?
Is it possible to find a map that can be used and completed (with create.go's `createOlmArtifactsForOCI, getImageDigests, renderDC, plan, etc)?
I was wondering if this would help reduce the burden of "rediscovering" the image contents at multiple places, sometimes through reading manifests, other times through walking the FS

Reply:
I took this idea and refactored the render code path to standardize on `map[OperatorCatalogPlatform]CatalogMetadata` so that we can capture new information in this map as the code goes through `getImageDigests` -> `renderDCFunc` -> `plan`. To do this, renderDCFunc now uses this map as an in/out parameter.